### ISSUE TITLE
Java 11 support in Gradle 4.10

### ISSUE THREAD

**Issue by joshuamillsma** [gradle/gradle#7835]
I forked gradle 4.10.2 locally in order to properly support java 11.  I noticed the following issues:
1. The code needs to use ASM7 (or ASM7 experimental) in order for test plugins to work properly.
2. jsr305 needs to be upgraded to 3.0.2
3. findbugs needs to be swapped out for spotbugs (essentially the same ASM issue).

I checked the master branch for gradle 5 and see that it already has these fixes incorporated, but switching to gradle 5 will be a non-trivial undertaking due to the changes in the publishing DSL.  Is gradle 4 a dead end at this point?  The changes are fairly minor and I can create a pull request.

**Comment by marcphilipp** [comment]
Which changes in the publishing DSL are you referring to?

**Comment by joshuamillsma** [comment]
Specifically, we get a warning about STABLE_PUBLISHING when publishing with
gradle 4:
https://docs.gradle.org/4.10.2/userguide/publishing_maven.html#publishing_maven:deferred_configuration

"Prior to Gradle 4.8, the publishing {} block was implicitly treated as if
all the logic inside it was executed after the project is evaluated. This
caused quite a bit of confusion, because it was the only block that behaved
that way. As part of the stabilization effort in Gradle 4.8, we are
deprecating this behavior and asking all users to migrate their build."

I will say that the deferred configuration behavior caused us quite a bit
of grief because we are modifying the default POM files generated by the
tasks in the maven-publish plugin.  We are using the "taskGraph.whenReady"
trick as a workaround (see
https://discuss.gradle.org/t/maven-publish-not-using-version-set-dynamically-in-taskgraph-whenready/19484
and
https://discuss.gradle.org/t/modifying-version-number-and-publishing-with-new-publishing-plugin/4788
for others who have mentioned this).  I am looking forward to ripping this
out when I have time, but I suspect it will not be trivial.

My company just switched from gradle 3 to gradle 4 a few months ago, and
that took a few weeks of work. I'd like to avoid having to do that again to
switch to gradle 5.  To add more complexity, we have a lot of projects in
play, and each of them independently declares which version of gradle it is
using.  Most, but not all, of the projects have been switched to gradle
4.10.  The remainder of the work is a bureaucratic issue, not a technical
one.  There are a few other "new feature added in gradle 4, old feature
deprecated in gradle 4, old feature removed in gradle 5" issues that we
papered over when we switched to gradle 4.10.  We have about a dozen
plugins that are shared across both gradle 3 and gradle 4 projects, so we
can't actually completely remove usage of the old features yet without
breaking the older projects.  In some cases, we were forced to add "if
gradle < 4 do one thing else do another" logic.  The publishing changes are
complicated enough that we don't want to use this strategy.

Long story short, if we can't get the patches to use ASM7 into gradle 4,
we'll probably just use our own fork until Q2 of next year.

**Comment by joshuamillsma** [comment]
[gradle-4.10.2.diffs.txt](https://github.com/gradle/gradle/files/2617493/gradle-4.10.2.diffs.txt)

**Comment by joshuamillsma** [comment]
95% of the diff was from me running sed on the source.  It looks like the gradle5 source moved the asm version into a separate constant, presumably to avoid this type of diff in the future.

**Comment by marcphilipp** [comment]
IIRC there were additional fixes required to make Gradle fully compatible with Java 11. I'll double-check.

**Comment by joshuamillsma** [comment]
That is certainly possible.  We only ran into issues with testng, junit, and findbugs.

**Comment by marcphilipp** [comment]
I'm afraid I have bad news. It's not just ASM but we also had to upgrade Spock (so we can run our tests on Java 11) and Groovy. As such it's too big a change for a minor release, not only in effort but also in potential negative impact.

Once you've enabled `STABLE_PUBLISHING` in 4.10.x, you can simply remove the flag from your settings script when upgrading to 5.0. If that causes any grief, feel free to open another issue or ask on the forum.

**Comment by joshuamillsma** [comment]
Ok, I forgot that I built gradle with java 10 rather than 11, presumably due to the breakage you pointed out.  It still *runs* fine with java 11.  We'll just stick with our forked gradle for the time being until gradle 5 becomes stable.

**Comment by joshuamillsma** [comment]
I ran into https://github.com/gradle/gradle/issues/6453, and the workaround in the ticket isn't working for me.  The commit is tagged for 5.1RC1.  What's the release date for that?  This leaves me back to either running my own fork, or running effectively a nightly build of unreleased software.
